Source/WebCore/dom/Document.cpp

```diff
@@ -8992,15 +8993,36 @@ void Document::handlePopoverLightDismiss(PointerEvent& event)
     RefPtr popoverToAvoidHiding = [&]() -> HTMLElement* {
         auto& target = downcast<Node>(*event.target());
         auto* startElement = is<Element>(target) ? &downcast<Element>(target) : target.parentElement();
-        for (auto& element : lineageOfType<HTMLElement>(*startElement)) {
-            if (element.popoverState() != PopoverState::Auto)
-                continue;
-            if (element.popoverData()->visibilityState() != PopoverVisibilityState::Showing)
-                continue;
-            return &element;
-        }
-        // FIXME: Handle popovertarget attributes.
-        return nullptr;
+        auto* clickedPopover = [&]() -> HTMLElement* {
+            for (auto& element : lineageOfType<HTMLElement>(*startElement)) {
+                if (element.popoverState() == PopoverState::Auto && element.popoverData()->visibilityState() == PopoverVisibilityState::Showing)
+                    return &element;
+            }
+            return nullptr;
+        }();
+
+        auto* invokerPopover = [&]() -> HTMLElement* {
+            for (auto& element : lineageOfType<HTMLFormControlElement>(*startElement)) {
+                auto* targetPopover = element.popoverTargetElement();
+                if (targetPopover && targetPopover->popoverState() == PopoverState::Auto && targetPopover->popoverData()->visibilityState() == PopoverVisibilityState::Showing)
+                    return targetPopover;
+            }
+            return nullptr;
+        }();
+
+        auto highestInTopLayer = [&](HTMLElement* first, HTMLElement* second) -> HTMLElement* {
+            if (!first)
+                return second;
+            if (!second)
+                return first;
+            Element* highestFound = nullptr;
+            for (auto& element : m_topLayerElements) {
```

Comment:
Wouldn't we want to iterate backwards so we don't have to traverse the whole ListHashSet?

Reply:
ah yes, very true